{
 "cells": [
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## 04c MPP bysurface differential analysis MAST full\n",
    "\n",
    "Celltype:\n",
    "\n",
    "**MPPs by surface markers (not clustered)**\n",
    "\n",
    "\n",
    "Run this model:\n",
    "\n",
    "`zlmCond_all <- zlm(formula = ~condition + Female + n_genes, sca=sca)`\n",
    "\n",
    "Comparisons:\n",
    "\n",
    "all cells\n",
    "- male vs female in all cells controlling for chemical treatment etc\n",
    "- chemical treatments controlling for sex and n_genes\n",
    "\n",
    "each cluster\n",
    "- male vs female in all cells controlling for chemical treatment etc\n",
    "- chemical treatments controlling for sex and n_genes\n",
    "\n",
    "\n",
    "done with this docker image:\n",
    "\n",
    "docker run --rm -d --name test_eva -p 8883:8888 -e JUPYTER_ENABLE_LAB=YES -v /Users/efast/Documents/:/home/jovyan/work r_scanpy:vs5\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdin",
     "output_type": "stream",
     "text": [
      "Once deleted, variables cannot be recovered. Proceed (y/[n])?  y\n"
     ]
    }
   ],
   "source": [
    "%reset"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import scanpy as sc\n",
    "import numpy as np\n",
    "import scipy as sp\n",
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "from matplotlib import rcParams\n",
    "from matplotlib import colors\n",
    "import seaborn as sb\n",
    "from gprofiler import GProfiler\n",
    "\n",
    "import rpy2.rinterface_lib.callbacks\n",
    "import logging\n",
    "\n",
    "from rpy2.robjects import pandas2ri\n",
    "import anndata2ri"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scanpy==1.4.5.1 anndata==0.7.1 umap==0.3.10 numpy==1.17.3 scipy==1.3.0 pandas==0.25.3 scikit-learn==0.22.2.post1 statsmodels==0.10.0 python-igraph==0.7.1 louvain==0.6.1\n"
     ]
    }
   ],
   "source": [
    "# Ignore R warning messages\n",
    "#Note: this can be commented out to get more verbose R output\n",
    "rpy2.rinterface_lib.callbacks.logger.setLevel(logging.ERROR)\n",
    "\n",
    "# Automatically convert rpy2 outputs to pandas dataframes\n",
    "pandas2ri.activate()\n",
    "anndata2ri.activate()\n",
    "%load_ext rpy2.ipython\n",
    "\n",
    "plt.rcParams['figure.figsize']=(8,8) #rescale figures\n",
    "sc.settings.verbosity = 3\n",
    "#sc.set_figure_params(dpi=200, dpi_save=300)\n",
    "sc.logging.print_versions()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "# Load libraries from correct lib Paths for my environment - ignore this!\n",
    ".libPaths(.libPaths()[c(3,2,1)])\n",
    "\n",
    "# Load all the R libraries we will be using in the notebook\n",
    "library(scran)\n",
    "library(ggplot2)\n",
    "library(plyr)\n",
    "library(MAST)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### MPPs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load data\n",
    "\n",
    "adata = sc.read(\n",
    "    './sc_objects/MAST_diffexpr_MPP_annotated.h5ad', cache = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create new Anndata object for use in MAST with non-batch corrected data as before\n",
    "adata_raw = adata.copy()\n",
    "adata_raw.X = adata.raw.X\n",
    "adata_raw.obs['n_genes'] = (adata_raw.X > 0).sum(1) # recompute number of genes expressed per cell\n",
    "adata = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 5819 × 13099 \n",
       "    obs: 'assignment', 'batch', 'counts', 'demux_type', 'hto_type', 'rna_type', 'sample', 'n_counts', 'log_counts', 'n_genes', 'percent_mito', 'Female', 'Female_cat', 'Female_str', 'sex_sample', 'rXist', 'leiden'\n",
       "    var: 'gene_ids', 'feature_types', 'n_cells', 'highly_variable', 'means', 'dispersions', 'dispersions_norm'\n",
       "    uns: 'leiden', 'leiden_colors', 'neighbors', 'pca', 'rank_genes_groups', 'sample_colors', 'sex_sample_colors', 'umap'\n",
       "    obsm: 'X_pca', 'X_umap'\n",
       "    varm: 'PCs'"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_raw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assignment</th>\n",
       "      <th>batch</th>\n",
       "      <th>counts</th>\n",
       "      <th>demux_type</th>\n",
       "      <th>hto_type</th>\n",
       "      <th>rna_type</th>\n",
       "      <th>sample</th>\n",
       "      <th>n_counts</th>\n",
       "      <th>log_counts</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>percent_mito</th>\n",
       "      <th>Female</th>\n",
       "      <th>Female_cat</th>\n",
       "      <th>Female_str</th>\n",
       "      <th>sex_sample</th>\n",
       "      <th>rXist</th>\n",
       "      <th>leiden</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCCAAGATGGGCT-0-0-0-0-0</th>\n",
       "      <td>MPP</td>\n",
       "      <td>batch1</td>\n",
       "      <td>1021.0</td>\n",
       "      <td>singlet</td>\n",
       "      <td>background</td>\n",
       "      <td>signal</td>\n",
       "      <td>ct</td>\n",
       "      <td>17629.0</td>\n",
       "      <td>9.777698</td>\n",
       "      <td>4802</td>\n",
       "      <td>0.049501</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>true</td>\n",
       "      <td>ct_true</td>\n",
       "      <td>3.338417</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAGGATAGTAGTCTC-0-0-0-0-0</th>\n",
       "      <td>MPP</td>\n",
       "      <td>batch1</td>\n",
       "      <td>809.0</td>\n",
       "      <td>singlet</td>\n",
       "      <td>background</td>\n",
       "      <td>signal</td>\n",
       "      <td>ct</td>\n",
       "      <td>13520.0</td>\n",
       "      <td>9.512590</td>\n",
       "      <td>4053</td>\n",
       "      <td>0.056176</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>false</td>\n",
       "      <td>ct_false</td>\n",
       "      <td>-0.119305</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAGGATCACGCTGAC-0-0-0-0-0</th>\n",
       "      <td>MPP</td>\n",
       "      <td>batch1</td>\n",
       "      <td>869.0</td>\n",
       "      <td>singlet</td>\n",
       "      <td>background</td>\n",
       "      <td>signal</td>\n",
       "      <td>ct</td>\n",
       "      <td>18169.0</td>\n",
       "      <td>9.807637</td>\n",
       "      <td>4619</td>\n",
       "      <td>0.049802</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>true</td>\n",
       "      <td>ct_true</td>\n",
       "      <td>3.066876</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAGGGCAGCAGCGAT-0-0-0-0-0</th>\n",
       "      <td>MPP</td>\n",
       "      <td>batch1</td>\n",
       "      <td>848.0</td>\n",
       "      <td>singlet</td>\n",
       "      <td>background</td>\n",
       "      <td>signal</td>\n",
       "      <td>ct</td>\n",
       "      <td>8510.0</td>\n",
       "      <td>9.049468</td>\n",
       "      <td>2986</td>\n",
       "      <td>0.056965</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>false</td>\n",
       "      <td>ct_false</td>\n",
       "      <td>-0.119305</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAGGTACATGAGATA-0-0-0-0-0</th>\n",
       "      <td>MPP</td>\n",
       "      <td>batch1</td>\n",
       "      <td>735.0</td>\n",
       "      <td>singlet</td>\n",
       "      <td>background</td>\n",
       "      <td>signal</td>\n",
       "      <td>ct</td>\n",
       "      <td>5256.0</td>\n",
       "      <td>8.567506</td>\n",
       "      <td>2038</td>\n",
       "      <td>0.057626</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>true</td>\n",
       "      <td>ct_true</td>\n",
       "      <td>2.669866</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           assignment   batch  counts demux_type    hto_type  \\\n",
       "AAACCCAAGATGGGCT-0-0-0-0-0        MPP  batch1  1021.0    singlet  background   \n",
       "AAAGGATAGTAGTCTC-0-0-0-0-0        MPP  batch1   809.0    singlet  background   \n",
       "AAAGGATCACGCTGAC-0-0-0-0-0        MPP  batch1   869.0    singlet  background   \n",
       "AAAGGGCAGCAGCGAT-0-0-0-0-0        MPP  batch1   848.0    singlet  background   \n",
       "AAAGGTACATGAGATA-0-0-0-0-0        MPP  batch1   735.0    singlet  background   \n",
       "\n",
       "                           rna_type sample  n_counts  log_counts  n_genes  \\\n",
       "AAACCCAAGATGGGCT-0-0-0-0-0   signal     ct   17629.0    9.777698     4802   \n",
       "AAAGGATAGTAGTCTC-0-0-0-0-0   signal     ct   13520.0    9.512590     4053   \n",
       "AAAGGATCACGCTGAC-0-0-0-0-0   signal     ct   18169.0    9.807637     4619   \n",
       "AAAGGGCAGCAGCGAT-0-0-0-0-0   signal     ct    8510.0    9.049468     2986   \n",
       "AAAGGTACATGAGATA-0-0-0-0-0   signal     ct    5256.0    8.567506     2038   \n",
       "\n",
       "                            percent_mito  Female Female_cat Female_str  \\\n",
       "AAACCCAAGATGGGCT-0-0-0-0-0      0.049501    True       True       true   \n",
       "AAAGGATAGTAGTCTC-0-0-0-0-0      0.056176   False      False      false   \n",
       "AAAGGATCACGCTGAC-0-0-0-0-0      0.049802    True       True       true   \n",
       "AAAGGGCAGCAGCGAT-0-0-0-0-0      0.056965   False      False      false   \n",
       "AAAGGTACATGAGATA-0-0-0-0-0      0.057626    True       True       true   \n",
       "\n",
       "                           sex_sample     rXist leiden  \n",
       "AAACCCAAGATGGGCT-0-0-0-0-0    ct_true  3.338417      4  \n",
       "AAAGGATAGTAGTCTC-0-0-0-0-0   ct_false -0.119305      1  \n",
       "AAAGGATCACGCTGAC-0-0-0-0-0    ct_true  3.066876      2  \n",
       "AAAGGGCAGCAGCGAT-0-0-0-0-0   ct_false -0.119305      2  \n",
       "AAAGGTACATGAGATA-0-0-0-0-0    ct_true  2.669866      0  "
      ]
     },
     "execution_count": 8,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_raw.obs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Select genes expressed in >5% of cells (no adaptive thresholding)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i adata_raw\n",
    "\n",
    "#Convert SingleCellExperiment to SingleCellAssay type as required by MAST\n",
    "sca <- SceToSingleCellAssay(adata_raw, class = \"SingleCellAssay\")\n",
    "\n",
    "#Scale Gene detection rate\n",
    "colData(sca)$n_genes = scale(colData(sca)$n_genes)\n",
    "\n",
    "# filter genes based on hard cutoff (have to be expressed in at least 5% of all cells)\n",
    "freq_expressed <- 0.05\n",
    "expressed_genes <- freq(sca) > freq_expressed\n",
    "sca <- sca[expressed_genes,]\n",
    "\n",
    "#rename the sample to condition and make the ct the control\n",
    "cond<-factor(colData(sca)$sample)\n",
    "cond<-relevel(cond,\"ct\")\n",
    "colData(sca)$condition<-cond"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### run MAST\n",
    "\n",
    "background:  \n",
    "`zlmCond_all <- zlm(formula = ~condition + Female + n_genes, sca=sca) # this runs the model`\n",
    "\n",
    "a formula with the measurement variable (gene expression) on the LHS (left hand side) and \n",
    "predictors present in colData on the RHS\n",
    "expression of genes controlling for cluster, condition, sex + n_genes\n",
    "questions I can ask:\n",
    "sex differences controlling for treatments\n",
    "sex differences controlling for clusters - not necessary analyze all the clusters\n",
    "overall gene expression changes in treatment\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 10,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R \n",
    "#Define & run hurdle model \n",
    "zlmCond_all <- zlm(formula = ~condition + Female + n_genes, sca=sca) # this runs the model\n",
    "summaryCond_all <- summary(zlmCond_all, doLRT=TRUE) # extracts the data, gives datatable with summary of fit, doLRT=TRUE extracts likelihood ratio test p-value\n",
    "summaryDt_all <- summaryCond_all$datatable # reformats into a table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       primerid component        contrast    Pr..Chisq.        ci.hi\n",
      "1 0610009B22Rik         C      FemaleTRUE  1.573119e-01  0.004737989\n",
      "2 0610009B22Rik         C   conditionGCSF  4.940772e-01  0.033506218\n",
      "3 0610009B22Rik         C conditiondmPGE2  2.041257e-02 -0.004391915\n",
      "4 0610009B22Rik         C   conditionindo  2.891810e-01  0.011299670\n",
      "5 0610009B22Rik         C    conditionpIC  2.261133e-05 -0.028569966\n",
      "6 0610009B22Rik         C         n_genes 1.494891e-249 -0.144555328\n",
      "        ci.lo         coef           z\n",
      "1 -0.02949859 -0.012380302  -1.4174865\n",
      "2 -0.01614560  0.008680311   0.6852961\n",
      "3 -0.05152801 -0.027959961  -2.3252041\n",
      "4 -0.03804252 -0.013371423  -1.0622759\n",
      "5 -0.07732979 -0.052949877  -4.2567774\n",
      "6 -0.15947825 -0.152016788 -39.9315195\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "head(summaryDt_all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -o female_all -o GCSF_all -o dmPGE2_all -o indo_all -o pIC_all\n",
    "\n",
    "# reformat for female\n",
    "result_all_Female <- merge(summaryDt_all[contrast=='FemaleTRUE' & component=='H',.(primerid, `Pr(>Chisq)`)], #P-vals\n",
    "                  summaryDt_all[contrast=='FemaleTRUE' & component=='logFC', .(primerid, coef)],\n",
    "                  by='primerid') #logFC coefficients\n",
    "#Correct for multiple testing (FDR correction) and filtering\n",
    "result_all_Female[,FDR:=p.adjust(`Pr(>Chisq)`, 'fdr')] # create column named FDR - probably that p.adjust function\n",
    "female_all = result_all_Female[result_all_Female$FDR<0.01,, drop=F] # create new table where rows with FDR<0.01 are droped\n",
    "female_all = female_all[order(female_all$FDR),] # sorts the table\n",
    "\n",
    "\n",
    "# reformat for GCSF\n",
    "result_all_GCSF <- merge(summaryDt_all[contrast=='conditionGCSF' & component=='H',.(primerid, `Pr(>Chisq)`)], #P-vals\n",
    "                  summaryDt_all[contrast=='conditionGCSF' & component=='logFC', .(primerid, coef)],\n",
    "                  by='primerid') #logFC coefficients\n",
    "#Correct for multiple testing (FDR correction) and filtering\n",
    "result_all_GCSF[,FDR:=p.adjust(`Pr(>Chisq)`, 'fdr')] # create column named FDR - probably that p.adjust function\n",
    "GCSF_all = result_all_GCSF[result_all_GCSF$FDR<0.01,, drop=F] # create new table where rows with FDR<0.01 are droped\n",
    "GCSF_all = GCSF_all[order(GCSF_all$FDR),] # sorts the table\n",
    "\n",
    "\n",
    "# reformat for dmPGE2\n",
    "result_all_dmPGE2 <- merge(summaryDt_all[contrast=='conditiondmPGE2' & component=='H',.(primerid, `Pr(>Chisq)`)], #P-vals\n",
    "                  summaryDt_all[contrast=='conditiondmPGE2' & component=='logFC', .(primerid, coef)],\n",
    "                  by='primerid') #logFC coefficients\n",
    "#Correct for multiple testing (FDR correction) and filtering\n",
    "result_all_dmPGE2[,FDR:=p.adjust(`Pr(>Chisq)`, 'fdr')] # create column named FDR - probably that p.adjust function\n",
    "dmPGE2_all = result_all_dmPGE2[result_all_dmPGE2$FDR<0.01,, drop=F] # create new table where rows with FDR<0.01 are droped\n",
    "dmPGE2_all = dmPGE2_all[order(dmPGE2_all$FDR),] # sorts the table\n",
    "\n",
    "\n",
    "# reformat for indo\n",
    "result_all_indo <- merge(summaryDt_all[contrast=='conditionindo' & component=='H',.(primerid, `Pr(>Chisq)`)], #P-vals\n",
    "                  summaryDt_all[contrast=='conditionindo' & component=='logFC', .(primerid, coef)],\n",
    "                  by='primerid') #logFC coefficients\n",
    "#Correct for multiple testing (FDR correction) and filtering\n",
    "result_all_indo[,FDR:=p.adjust(`Pr(>Chisq)`, 'fdr')] # create column named FDR - probably that p.adjust function\n",
    "indo_all = result_all_indo[result_all_indo$FDR<0.01,, drop=F] # create new table where rows with FDR<0.01 are droped\n",
    "indo_all = indo_all[order(indo_all$FDR),] # sorts the table\n",
    "\n",
    "# reformat for pIC\n",
    "result_all_pIC <- merge(summaryDt_all[contrast=='conditionpIC' & component=='H',.(primerid, `Pr(>Chisq)`)], #P-vals\n",
    "                  summaryDt_all[contrast=='conditionpIC' & component=='logFC', .(primerid, coef)],\n",
    "                  by='primerid') #logFC coefficients\n",
    "#Correct for multiple testing (FDR correction) and filtering\n",
    "result_all_pIC[,FDR:=p.adjust(`Pr(>Chisq)`, 'fdr')] # create column named FDR - probably that p.adjust function\n",
    "pIC_all = result_all_pIC[result_all_pIC$FDR<0.01,, drop=F] # create new table where rows with FDR<0.01 are droped\n",
    "pIC_all = pIC_all[order(pIC_all$FDR),] # sorts the table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 13,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -o MAST_raw_all\n",
    "\n",
    "MAST_raw_all <- summaryDt_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 14,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save files as .csvs\n",
    "\n",
    "MAST_raw_all.to_csv('./write/MPP_MAST_raw_surf.csv')\n",
    "female_all.to_csv('./write/MPP_MAST_female_surf.csv')\n",
    "GCSF_all.to_csv('./write/MPP_MAST_GCSF_surf.csv')\n",
    "pIC_all.to_csv('./write/MPP_MAST_pIC_surf.csv')\n",
    "dmPGE2_all.to_csv('./write/MPP_MAST_dmPGE2_surf.csv')\n",
    "indo_all.to_csv('./write/MPP_MAST_indo_surf.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove python variables\n",
    "\n",
    "adata_raw = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "# remove R variables\n",
    "\n",
    "rm(adata_raw)\n",
    "rm(zlmCond_all)\n",
    "rm(summaryDt_all)\n",
    "rm(summaryCond_all)\n",
    "rm(MAST_raw_all)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### MPP1s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load data\n",
    "\n",
    "adata = sc.read(\n",
    "    './sc_objects/MAST_diffexpr_MPP1_annotated.h5ad', cache = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create new Anndata object for use in MAST with non-batch corrected data as before\n",
    "adata_raw = adata.copy()\n",
    "adata_raw.X = adata.raw.X\n",
    "adata_raw.obs['n_genes'] = (adata_raw.X > 0).sum(1) # recompute number of genes expressed per cell\n",
    "adata = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 5343 × 13286 \n",
       "    obs: 'assignment', 'batch', 'counts', 'demux_type', 'hto_type', 'rna_type', 'sample', 'n_counts', 'log_counts', 'n_genes', 'percent_mito', 'Female', 'Female_cat', 'Female_str', 'sex_sample', 'rXist', 'leiden'\n",
       "    var: 'gene_ids', 'feature_types', 'n_cells', 'highly_variable', 'means', 'dispersions', 'dispersions_norm'\n",
       "    uns: 'leiden', 'leiden_colors', 'neighbors', 'pca', 'rank_genes_groups', 'sample_colors', 'sex_sample_colors', 'umap'\n",
       "    obsm: 'X_pca', 'X_umap'\n",
       "    varm: 'PCs'"
      ]
     },
     "execution_count": 19,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_raw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assignment</th>\n",
       "      <th>batch</th>\n",
       "      <th>counts</th>\n",
       "      <th>demux_type</th>\n",
       "      <th>hto_type</th>\n",
       "      <th>rna_type</th>\n",
       "      <th>sample</th>\n",
       "      <th>n_counts</th>\n",
       "      <th>log_counts</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>percent_mito</th>\n",
       "      <th>Female</th>\n",
       "      <th>Female_cat</th>\n",
       "      <th>Female_str</th>\n",
       "      <th>sex_sample</th>\n",
       "      <th>rXist</th>\n",
       "      <th>leiden</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACCCATCACATTGG-0-0-0-0-0</th>\n",
       "      <td>MPP1</td>\n",
       "      <td>batch1</td>\n",
       "      <td>781.0</td>\n",
       "      <td>singlet</td>\n",
       "      <td>background</td>\n",
       "      <td>signal</td>\n",
       "      <td>ct</td>\n",
       "      <td>7089.0</td>\n",
       "      <td>8.866864</td>\n",
       "      <td>2545</td>\n",
       "      <td>0.063725</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>false</td>\n",
       "      <td>ct_false</td>\n",
       "      <td>0.006942</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACGAAGTACTGGGA-0-0-0-0-0</th>\n",
       "      <td>MPP1</td>\n",
       "      <td>batch1</td>\n",
       "      <td>835.0</td>\n",
       "      <td>singlet</td>\n",
       "      <td>background</td>\n",
       "      <td>signal</td>\n",
       "      <td>ct</td>\n",
       "      <td>10922.0</td>\n",
       "      <td>9.298627</td>\n",
       "      <td>3284</td>\n",
       "      <td>0.055022</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>true</td>\n",
       "      <td>ct_true</td>\n",
       "      <td>2.952886</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAACGAATCTCGGTAA-0-0-0-0-0</th>\n",
       "      <td>MPP1</td>\n",
       "      <td>batch1</td>\n",
       "      <td>745.0</td>\n",
       "      <td>singlet</td>\n",
       "      <td>background</td>\n",
       "      <td>signal</td>\n",
       "      <td>ct</td>\n",
       "      <td>7894.0</td>\n",
       "      <td>8.974238</td>\n",
       "      <td>2692</td>\n",
       "      <td>0.050905</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>true</td>\n",
       "      <td>ct_true</td>\n",
       "      <td>2.404951</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAGAACTCGACCATA-0-0-0-0-0</th>\n",
       "      <td>MPP1</td>\n",
       "      <td>batch1</td>\n",
       "      <td>1210.0</td>\n",
       "      <td>singlet</td>\n",
       "      <td>signal</td>\n",
       "      <td>signal</td>\n",
       "      <td>ct</td>\n",
       "      <td>19285.0</td>\n",
       "      <td>9.867290</td>\n",
       "      <td>4997</td>\n",
       "      <td>0.044844</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>false</td>\n",
       "      <td>ct_false</td>\n",
       "      <td>0.006942</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAGGATAGAAGGGAT-0-0-0-0-0</th>\n",
       "      <td>MPP1</td>\n",
       "      <td>batch1</td>\n",
       "      <td>701.0</td>\n",
       "      <td>singlet</td>\n",
       "      <td>background</td>\n",
       "      <td>signal</td>\n",
       "      <td>ct</td>\n",
       "      <td>6741.0</td>\n",
       "      <td>8.816854</td>\n",
       "      <td>2460</td>\n",
       "      <td>0.073811</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>false</td>\n",
       "      <td>ct_false</td>\n",
       "      <td>0.006942</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           assignment   batch  counts demux_type    hto_type  \\\n",
       "AAACCCATCACATTGG-0-0-0-0-0       MPP1  batch1   781.0    singlet  background   \n",
       "AAACGAAGTACTGGGA-0-0-0-0-0       MPP1  batch1   835.0    singlet  background   \n",
       "AAACGAATCTCGGTAA-0-0-0-0-0       MPP1  batch1   745.0    singlet  background   \n",
       "AAAGAACTCGACCATA-0-0-0-0-0       MPP1  batch1  1210.0    singlet      signal   \n",
       "AAAGGATAGAAGGGAT-0-0-0-0-0       MPP1  batch1   701.0    singlet  background   \n",
       "\n",
       "                           rna_type sample  n_counts  log_counts  n_genes  \\\n",
       "AAACCCATCACATTGG-0-0-0-0-0   signal     ct    7089.0    8.866864     2545   \n",
       "AAACGAAGTACTGGGA-0-0-0-0-0   signal     ct   10922.0    9.298627     3284   \n",
       "AAACGAATCTCGGTAA-0-0-0-0-0   signal     ct    7894.0    8.974238     2692   \n",
       "AAAGAACTCGACCATA-0-0-0-0-0   signal     ct   19285.0    9.867290     4997   \n",
       "AAAGGATAGAAGGGAT-0-0-0-0-0   signal     ct    6741.0    8.816854     2460   \n",
       "\n",
       "                            percent_mito  Female Female_cat Female_str  \\\n",
       "AAACCCATCACATTGG-0-0-0-0-0      0.063725   False      False      false   \n",
       "AAACGAAGTACTGGGA-0-0-0-0-0      0.055022    True       True       true   \n",
       "AAACGAATCTCGGTAA-0-0-0-0-0      0.050905    True       True       true   \n",
       "AAAGAACTCGACCATA-0-0-0-0-0      0.044844   False      False      false   \n",
       "AAAGGATAGAAGGGAT-0-0-0-0-0      0.073811   False      False      false   \n",
       "\n",
       "                           sex_sample     rXist leiden  \n",
       "AAACCCATCACATTGG-0-0-0-0-0   ct_false  0.006942      0  \n",
       "AAACGAAGTACTGGGA-0-0-0-0-0    ct_true  2.952886      1  \n",
       "AAACGAATCTCGGTAA-0-0-0-0-0    ct_true  2.404951      0  \n",
       "AAAGAACTCGACCATA-0-0-0-0-0   ct_false  0.006942      1  \n",
       "AAAGGATAGAAGGGAT-0-0-0-0-0   ct_false  0.006942      0  "
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_raw.obs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Select genes expressed in >5% of cells (no adaptive thresholding)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i adata_raw\n",
    "\n",
    "#Convert SingleCellExperiment to SingleCellAssay type as required by MAST\n",
    "sca <- SceToSingleCellAssay(adata_raw, class = \"SingleCellAssay\")\n",
    "\n",
    "#Scale Gene detection rate\n",
    "colData(sca)$n_genes = scale(colData(sca)$n_genes)\n",
    "\n",
    "# filter genes based on hard cutoff (have to be expressed in at least 5% of all cells)\n",
    "freq_expressed <- 0.05\n",
    "expressed_genes <- freq(sca) > freq_expressed\n",
    "sca <- sca[expressed_genes,]\n",
    "\n",
    "#rename the sample to condition and make the ct the control\n",
    "cond<-factor(colData(sca)$sample)\n",
    "cond<-relevel(cond,\"ct\")\n",
    "colData(sca)$condition<-cond"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### run MAST\n",
    "\n",
    "background:  \n",
    "`zlmCond_all <- zlm(formula = ~condition + Female + n_genes, sca=sca) # this runs the model`\n",
    "\n",
    "a formula with the measurement variable (gene expression) on the LHS (left hand side) and \n",
    "predictors present in colData on the RHS\n",
    "expression of genes controlling for cluster, condition, sex + n_genes\n",
    "questions I can ask:\n",
    "sex differences controlling for treatments\n",
    "sex differences controlling for clusters - not necessary analyze all the clusters\n",
    "overall gene expression changes in treatment\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R \n",
    "#Define & run hurdle model \n",
    "zlmCond_all <- zlm(formula = ~condition + Female + n_genes, sca=sca) # this runs the model\n",
    "summaryCond_all <- summary(zlmCond_all, doLRT=TRUE) # extracts the data, gives datatable with summary of fit, doLRT=TRUE extracts likelihood ratio test p-value\n",
    "summaryDt_all <- summaryCond_all$datatable # reformats into a table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       primerid component        contrast    Pr..Chisq.        ci.hi\n",
      "1 0610009B22Rik         C      FemaleTRUE  1.630397e-01  0.004565518\n",
      "2 0610009B22Rik         C   conditionGCSF  8.439253e-01  0.024484451\n",
      "3 0610009B22Rik         C conditiondmPGE2  4.531509e-01  0.021603937\n",
      "4 0610009B22Rik         C   conditionindo  1.882372e-01  0.006679173\n",
      "5 0610009B22Rik         C    conditionpIC  9.056294e-04 -0.016493551\n",
      "6 0610009B22Rik         C         n_genes 6.408240e-195 -0.120319223\n",
      "        ci.lo         coef           z\n",
      "1 -0.02722212 -0.011328300  -1.3969620\n",
      "2 -0.02995997 -0.002737758  -0.1971151\n",
      "3 -0.04845231 -0.013424189  -0.7511372\n",
      "4 -0.03408587 -0.013703349  -1.3177011\n",
      "5 -0.06378743 -0.040140493  -3.3270231\n",
      "6 -0.13525281 -0.127786014 -33.5426520\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "head(summaryDt_all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -o female_all -o GCSF_all -o dmPGE2_all -o indo_all -o pIC_all\n",
    "\n",
    "# reformat for female\n",
    "result_all_Female <- merge(summaryDt_all[contrast=='FemaleTRUE' & component=='H',.(primerid, `Pr(>Chisq)`)], #P-vals\n",
    "                  summaryDt_all[contrast=='FemaleTRUE' & component=='logFC', .(primerid, coef)],\n",
    "                  by='primerid') #logFC coefficients\n",
    "#Correct for multiple testing (FDR correction) and filtering\n",
    "result_all_Female[,FDR:=p.adjust(`Pr(>Chisq)`, 'fdr')] # create column named FDR - probably that p.adjust function\n",
    "female_all = result_all_Female[result_all_Female$FDR<0.01,, drop=F] # create new table where rows with FDR<0.01 are droped\n",
    "female_all = female_all[order(female_all$FDR),] # sorts the table\n",
    "\n",
    "\n",
    "# reformat for GCSF\n",
    "result_all_GCSF <- merge(summaryDt_all[contrast=='conditionGCSF' & component=='H',.(primerid, `Pr(>Chisq)`)], #P-vals\n",
    "                  summaryDt_all[contrast=='conditionGCSF' & component=='logFC', .(primerid, coef)],\n",
    "                  by='primerid') #logFC coefficients\n",
    "#Correct for multiple testing (FDR correction) and filtering\n",
    "result_all_GCSF[,FDR:=p.adjust(`Pr(>Chisq)`, 'fdr')] # create column named FDR - probably that p.adjust function\n",
    "GCSF_all = result_all_GCSF[result_all_GCSF$FDR<0.01,, drop=F] # create new table where rows with FDR<0.01 are droped\n",
    "GCSF_all = GCSF_all[order(GCSF_all$FDR),] # sorts the table\n",
    "\n",
    "\n",
    "# reformat for dmPGE2\n",
    "result_all_dmPGE2 <- merge(summaryDt_all[contrast=='conditiondmPGE2' & component=='H',.(primerid, `Pr(>Chisq)`)], #P-vals\n",
    "                  summaryDt_all[contrast=='conditiondmPGE2' & component=='logFC', .(primerid, coef)],\n",
    "                  by='primerid') #logFC coefficients\n",
    "#Correct for multiple testing (FDR correction) and filtering\n",
    "result_all_dmPGE2[,FDR:=p.adjust(`Pr(>Chisq)`, 'fdr')] # create column named FDR - probably that p.adjust function\n",
    "dmPGE2_all = result_all_dmPGE2[result_all_dmPGE2$FDR<0.01,, drop=F] # create new table where rows with FDR<0.01 are droped\n",
    "dmPGE2_all = dmPGE2_all[order(dmPGE2_all$FDR),] # sorts the table\n",
    "\n",
    "\n",
    "# reformat for indo\n",
    "result_all_indo <- merge(summaryDt_all[contrast=='conditionindo' & component=='H',.(primerid, `Pr(>Chisq)`)], #P-vals\n",
    "                  summaryDt_all[contrast=='conditionindo' & component=='logFC', .(primerid, coef)],\n",
    "                  by='primerid') #logFC coefficients\n",
    "#Correct for multiple testing (FDR correction) and filtering\n",
    "result_all_indo[,FDR:=p.adjust(`Pr(>Chisq)`, 'fdr')] # create column named FDR - probably that p.adjust function\n",
    "indo_all = result_all_indo[result_all_indo$FDR<0.01,, drop=F] # create new table where rows with FDR<0.01 are droped\n",
    "indo_all = indo_all[order(indo_all$FDR),] # sorts the table\n",
    "\n",
    "# reformat for pIC\n",
    "result_all_pIC <- merge(summaryDt_all[contrast=='conditionpIC' & component=='H',.(primerid, `Pr(>Chisq)`)], #P-vals\n",
    "                  summaryDt_all[contrast=='conditionpIC' & component=='logFC', .(primerid, coef)],\n",
    "                  by='primerid') #logFC coefficients\n",
    "#Correct for multiple testing (FDR correction) and filtering\n",
    "result_all_pIC[,FDR:=p.adjust(`Pr(>Chisq)`, 'fdr')] # create column named FDR - probably that p.adjust function\n",
    "pIC_all = result_all_pIC[result_all_pIC$FDR<0.01,, drop=F] # create new table where rows with FDR<0.01 are droped\n",
    "pIC_all = pIC_all[order(pIC_all$FDR),] # sorts the table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -o MAST_raw_all\n",
    "\n",
    "MAST_raw_all <- summaryDt_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save files as .csvs\n",
    "\n",
    "MAST_raw_all.to_csv('./write/MPP1_MAST_raw_surf.csv')\n",
    "female_all.to_csv('./write/MPP1_MAST_female_surf.csv')\n",
    "GCSF_all.to_csv('./write/MPP1_MAST_GCSF_surf.csv')\n",
    "pIC_all.to_csv('./write/MPP1_MAST_pIC_surf.csv')\n",
    "dmPGE2_all.to_csv('./write/MPP1_MAST_dmPGE2_surf.csv')\n",
    "indo_all.to_csv('./write/MPP1_MAST_indo_surf.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove python variables\n",
    "\n",
    "adata_raw = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "# remove R variables\n",
    "\n",
    "rm(adata_raw)\n",
    "rm(zlmCond_all)\n",
    "rm(summaryDt_all)\n",
    "rm(summaryCond_all)\n",
    "rm(MAST_raw_all)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### MPP2s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load data\n",
    "\n",
    "adata = sc.read(\n",
    "    './sc_objects/MAST_diffexpr_MPP2_annotated.h5ad', cache = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create new Anndata object for use in MAST with non-batch corrected data as before\n",
    "adata_raw = adata.copy()\n",
    "adata_raw.X = adata.raw.X\n",
    "adata_raw.obs['n_genes'] = (adata_raw.X > 0).sum(1) # recompute number of genes expressed per cell\n",
    "adata = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 5884 × 13472 \n",
       "    obs: 'assignment', 'batch', 'counts', 'demux_type', 'hto_type', 'rna_type', 'sample', 'n_counts', 'log_counts', 'n_genes', 'percent_mito', 'Female', 'Female_cat', 'Female_str', 'sex_sample', 'rXist', 'leiden'\n",
       "    var: 'gene_ids', 'feature_types', 'n_cells', 'highly_variable', 'means', 'dispersions', 'dispersions_norm'\n",
       "    uns: 'leiden', 'leiden_colors', 'neighbors', 'pca', 'rank_genes_groups', 'sample_colors', 'sex_sample_colors', 'umap'\n",
       "    obsm: 'X_pca', 'X_umap'\n",
       "    varm: 'PCs'"
      ]
     },
     "execution_count": 31,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_raw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assignment</th>\n",
       "      <th>batch</th>\n",
       "      <th>counts</th>\n",
       "      <th>demux_type</th>\n",
       "      <th>hto_type</th>\n",
       "      <th>rna_type</th>\n",
       "      <th>sample</th>\n",
       "      <th>n_counts</th>\n",
       "      <th>log_counts</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>percent_mito</th>\n",
       "      <th>Female</th>\n",
       "      <th>Female_cat</th>\n",
       "      <th>Female_str</th>\n",
       "      <th>sex_sample</th>\n",
       "      <th>rXist</th>\n",
       "      <th>leiden</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAAGAACAGGAAGTCC-0-0-0-0-0</th>\n",
       "      <td>MPP2</td>\n",
       "      <td>batch1</td>\n",
       "      <td>1180.0</td>\n",
       "      <td>singlet</td>\n",
       "      <td>signal</td>\n",
       "      <td>signal</td>\n",
       "      <td>ct</td>\n",
       "      <td>22280.0</td>\n",
       "      <td>10.011580</td>\n",
       "      <td>5458</td>\n",
       "      <td>0.051564</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>true</td>\n",
       "      <td>ct_true</td>\n",
       "      <td>3.124767</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAGAACGTAACCCTA-0-0-0-0-0</th>\n",
       "      <td>MPP2</td>\n",
       "      <td>batch1</td>\n",
       "      <td>975.0</td>\n",
       "      <td>singlet</td>\n",
       "      <td>background</td>\n",
       "      <td>signal</td>\n",
       "      <td>ct</td>\n",
       "      <td>24253.0</td>\n",
       "      <td>10.096501</td>\n",
       "      <td>5472</td>\n",
       "      <td>0.050004</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>false</td>\n",
       "      <td>ct_false</td>\n",
       "      <td>0.178026</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAGGATGTACACGCC-0-0-0-0-0</th>\n",
       "      <td>MPP2</td>\n",
       "      <td>batch1</td>\n",
       "      <td>805.0</td>\n",
       "      <td>singlet</td>\n",
       "      <td>background</td>\n",
       "      <td>signal</td>\n",
       "      <td>ct</td>\n",
       "      <td>19423.0</td>\n",
       "      <td>9.874368</td>\n",
       "      <td>4714</td>\n",
       "      <td>0.036909</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>true</td>\n",
       "      <td>ct_true</td>\n",
       "      <td>3.006543</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAGGATGTACGAGTG-0-0-0-0-0</th>\n",
       "      <td>MPP2</td>\n",
       "      <td>batch1</td>\n",
       "      <td>810.0</td>\n",
       "      <td>singlet</td>\n",
       "      <td>background</td>\n",
       "      <td>signal</td>\n",
       "      <td>ct</td>\n",
       "      <td>8478.0</td>\n",
       "      <td>9.046056</td>\n",
       "      <td>2916</td>\n",
       "      <td>0.055863</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>true</td>\n",
       "      <td>ct_true</td>\n",
       "      <td>3.358304</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAGGATGTCATCCGG-0-0-0-0-0</th>\n",
       "      <td>MPP2</td>\n",
       "      <td>batch1</td>\n",
       "      <td>467.0</td>\n",
       "      <td>singlet</td>\n",
       "      <td>background</td>\n",
       "      <td>signal</td>\n",
       "      <td>ct</td>\n",
       "      <td>15443.0</td>\n",
       "      <td>9.645170</td>\n",
       "      <td>4524</td>\n",
       "      <td>0.045510</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>true</td>\n",
       "      <td>ct_true</td>\n",
       "      <td>3.327009</td>\n",
       "      <td>4</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           assignment   batch  counts demux_type    hto_type  \\\n",
       "AAAGAACAGGAAGTCC-0-0-0-0-0       MPP2  batch1  1180.0    singlet      signal   \n",
       "AAAGAACGTAACCCTA-0-0-0-0-0       MPP2  batch1   975.0    singlet  background   \n",
       "AAAGGATGTACACGCC-0-0-0-0-0       MPP2  batch1   805.0    singlet  background   \n",
       "AAAGGATGTACGAGTG-0-0-0-0-0       MPP2  batch1   810.0    singlet  background   \n",
       "AAAGGATGTCATCCGG-0-0-0-0-0       MPP2  batch1   467.0    singlet  background   \n",
       "\n",
       "                           rna_type sample  n_counts  log_counts  n_genes  \\\n",
       "AAAGAACAGGAAGTCC-0-0-0-0-0   signal     ct   22280.0   10.011580     5458   \n",
       "AAAGAACGTAACCCTA-0-0-0-0-0   signal     ct   24253.0   10.096501     5472   \n",
       "AAAGGATGTACACGCC-0-0-0-0-0   signal     ct   19423.0    9.874368     4714   \n",
       "AAAGGATGTACGAGTG-0-0-0-0-0   signal     ct    8478.0    9.046056     2916   \n",
       "AAAGGATGTCATCCGG-0-0-0-0-0   signal     ct   15443.0    9.645170     4524   \n",
       "\n",
       "                            percent_mito  Female Female_cat Female_str  \\\n",
       "AAAGAACAGGAAGTCC-0-0-0-0-0      0.051564    True       True       true   \n",
       "AAAGAACGTAACCCTA-0-0-0-0-0      0.050004   False      False      false   \n",
       "AAAGGATGTACACGCC-0-0-0-0-0      0.036909    True       True       true   \n",
       "AAAGGATGTACGAGTG-0-0-0-0-0      0.055863    True       True       true   \n",
       "AAAGGATGTCATCCGG-0-0-0-0-0      0.045510    True       True       true   \n",
       "\n",
       "                           sex_sample     rXist leiden  \n",
       "AAAGAACAGGAAGTCC-0-0-0-0-0    ct_true  3.124767      3  \n",
       "AAAGAACGTAACCCTA-0-0-0-0-0   ct_false  0.178026      4  \n",
       "AAAGGATGTACACGCC-0-0-0-0-0    ct_true  3.006543      3  \n",
       "AAAGGATGTACGAGTG-0-0-0-0-0    ct_true  3.358304      0  \n",
       "AAAGGATGTCATCCGG-0-0-0-0-0    ct_true  3.327009      4  "
      ]
     },
     "execution_count": 32,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_raw.obs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Select genes expressed in >5% of cells (no adaptive thresholding)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i adata_raw\n",
    "\n",
    "#Convert SingleCellExperiment to SingleCellAssay type as required by MAST\n",
    "sca <- SceToSingleCellAssay(adata_raw, class = \"SingleCellAssay\")\n",
    "\n",
    "#Scale Gene detection rate\n",
    "colData(sca)$n_genes = scale(colData(sca)$n_genes)\n",
    "\n",
    "# filter genes based on hard cutoff (have to be expressed in at least 5% of all cells)\n",
    "freq_expressed <- 0.05\n",
    "expressed_genes <- freq(sca) > freq_expressed\n",
    "sca <- sca[expressed_genes,]\n",
    "\n",
    "#rename the sample to condition and make the ct the control\n",
    "cond<-factor(colData(sca)$sample)\n",
    "cond<-relevel(cond,\"ct\")\n",
    "colData(sca)$condition<-cond"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### run MAST\n",
    "\n",
    "background:  \n",
    "`zlmCond_all <- zlm(formula = ~condition + Female + n_genes, sca=sca) # this runs the model`\n",
    "\n",
    "a formula with the measurement variable (gene expression) on the LHS (left hand side) and \n",
    "predictors present in colData on the RHS\n",
    "expression of genes controlling for cluster, condition, sex + n_genes\n",
    "questions I can ask:\n",
    "sex differences controlling for treatments\n",
    "sex differences controlling for clusters - not necessary analyze all the clusters\n",
    "overall gene expression changes in treatment\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 34,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R \n",
    "#Define & run hurdle model \n",
    "zlmCond_all <- zlm(formula = ~condition + Female + n_genes, sca=sca) # this runs the model\n",
    "summaryCond_all <- summary(zlmCond_all, doLRT=TRUE) # extracts the data, gives datatable with summary of fit, doLRT=TRUE extracts likelihood ratio test p-value\n",
    "summaryDt_all <- summaryCond_all$datatable # reformats into a table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 35,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       primerid component        contrast    Pr..Chisq.        ci.hi\n",
      "1 0610009B22Rik         C      FemaleTRUE  6.373351e-03 -0.005485245\n",
      "2 0610009B22Rik         C   conditionGCSF  7.164278e-01  0.015075893\n",
      "3 0610009B22Rik         C conditiondmPGE2  1.536900e-07 -0.032350728\n",
      "4 0610009B22Rik         C   conditionindo  5.259589e-01  0.015972727\n",
      "5 0610009B22Rik         C    conditionpIC  4.072594e-01  0.013843542\n",
      "6 0610009B22Rik         C         n_genes 1.316541e-197 -0.121172273\n",
      "        ci.lo         coef           z\n",
      "1 -0.03339868 -0.019441961  -2.7302657\n",
      "2 -0.02193617 -0.003430139  -0.3632842\n",
      "3 -0.07074231 -0.051546517  -5.2630980\n",
      "4 -0.03125741 -0.007642340  -0.6342862\n",
      "5 -0.03413314 -0.010144799  -0.8288794\n",
      "6 -0.13653095 -0.128851611 -32.8862368\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "head(summaryDt_all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -o female_all -o GCSF_all -o dmPGE2_all -o indo_all -o pIC_all\n",
    "\n",
    "# reformat for female\n",
    "result_all_Female <- merge(summaryDt_all[contrast=='FemaleTRUE' & component=='H',.(primerid, `Pr(>Chisq)`)], #P-vals\n",
    "                  summaryDt_all[contrast=='FemaleTRUE' & component=='logFC', .(primerid, coef)],\n",
    "                  by='primerid') #logFC coefficients\n",
    "#Correct for multiple testing (FDR correction) and filtering\n",
    "result_all_Female[,FDR:=p.adjust(`Pr(>Chisq)`, 'fdr')] # create column named FDR - probably that p.adjust function\n",
    "female_all = result_all_Female[result_all_Female$FDR<0.01,, drop=F] # create new table where rows with FDR<0.01 are droped\n",
    "female_all = female_all[order(female_all$FDR),] # sorts the table\n",
    "\n",
    "\n",
    "# reformat for GCSF\n",
    "result_all_GCSF <- merge(summaryDt_all[contrast=='conditionGCSF' & component=='H',.(primerid, `Pr(>Chisq)`)], #P-vals\n",
    "                  summaryDt_all[contrast=='conditionGCSF' & component=='logFC', .(primerid, coef)],\n",
    "                  by='primerid') #logFC coefficients\n",
    "#Correct for multiple testing (FDR correction) and filtering\n",
    "result_all_GCSF[,FDR:=p.adjust(`Pr(>Chisq)`, 'fdr')] # create column named FDR - probably that p.adjust function\n",
    "GCSF_all = result_all_GCSF[result_all_GCSF$FDR<0.01,, drop=F] # create new table where rows with FDR<0.01 are droped\n",
    "GCSF_all = GCSF_all[order(GCSF_all$FDR),] # sorts the table\n",
    "\n",
    "\n",
    "# reformat for dmPGE2\n",
    "result_all_dmPGE2 <- merge(summaryDt_all[contrast=='conditiondmPGE2' & component=='H',.(primerid, `Pr(>Chisq)`)], #P-vals\n",
    "                  summaryDt_all[contrast=='conditiondmPGE2' & component=='logFC', .(primerid, coef)],\n",
    "                  by='primerid') #logFC coefficients\n",
    "#Correct for multiple testing (FDR correction) and filtering\n",
    "result_all_dmPGE2[,FDR:=p.adjust(`Pr(>Chisq)`, 'fdr')] # create column named FDR - probably that p.adjust function\n",
    "dmPGE2_all = result_all_dmPGE2[result_all_dmPGE2$FDR<0.01,, drop=F] # create new table where rows with FDR<0.01 are droped\n",
    "dmPGE2_all = dmPGE2_all[order(dmPGE2_all$FDR),] # sorts the table\n",
    "\n",
    "\n",
    "# reformat for indo\n",
    "result_all_indo <- merge(summaryDt_all[contrast=='conditionindo' & component=='H',.(primerid, `Pr(>Chisq)`)], #P-vals\n",
    "                  summaryDt_all[contrast=='conditionindo' & component=='logFC', .(primerid, coef)],\n",
    "                  by='primerid') #logFC coefficients\n",
    "#Correct for multiple testing (FDR correction) and filtering\n",
    "result_all_indo[,FDR:=p.adjust(`Pr(>Chisq)`, 'fdr')] # create column named FDR - probably that p.adjust function\n",
    "indo_all = result_all_indo[result_all_indo$FDR<0.01,, drop=F] # create new table where rows with FDR<0.01 are droped\n",
    "indo_all = indo_all[order(indo_all$FDR),] # sorts the table\n",
    "\n",
    "# reformat for pIC\n",
    "result_all_pIC <- merge(summaryDt_all[contrast=='conditionpIC' & component=='H',.(primerid, `Pr(>Chisq)`)], #P-vals\n",
    "                  summaryDt_all[contrast=='conditionpIC' & component=='logFC', .(primerid, coef)],\n",
    "                  by='primerid') #logFC coefficients\n",
    "#Correct for multiple testing (FDR correction) and filtering\n",
    "result_all_pIC[,FDR:=p.adjust(`Pr(>Chisq)`, 'fdr')] # create column named FDR - probably that p.adjust function\n",
    "pIC_all = result_all_pIC[result_all_pIC$FDR<0.01,, drop=F] # create new table where rows with FDR<0.01 are droped\n",
    "pIC_all = pIC_all[order(pIC_all$FDR),] # sorts the table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -o MAST_raw_all\n",
    "\n",
    "MAST_raw_all <- summaryDt_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save files as .csvs\n",
    "\n",
    "MAST_raw_all.to_csv('./write/MPP2_MAST_raw_surf.csv')\n",
    "female_all.to_csv('./write/MPP2_MAST_female_surf.csv')\n",
    "GCSF_all.to_csv('./write/MPP2_MAST_GCSF_surf.csv')\n",
    "pIC_all.to_csv('./write/MPP2_MAST_pIC_surf.csv')\n",
    "dmPGE2_all.to_csv('./write/MPP2_MAST_dmPGE2_surf.csv')\n",
    "indo_all.to_csv('./write/MPP2_MAST_indo_surf.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove python variables\n",
    "\n",
    "adata_raw = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "# remove R variables\n",
    "\n",
    "rm(adata_raw)\n",
    "rm(zlmCond_all)\n",
    "rm(summaryDt_all)\n",
    "rm(summaryCond_all)\n",
    "rm(MAST_raw_all)"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### MPP3s"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [],
   "source": [
    "# load data\n",
    "\n",
    "adata = sc.read(\n",
    "    './sc_objects/MAST_diffexpr_MPP3_annotated.h5ad', cache = True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Create new Anndata object for use in MAST with non-batch corrected data as before\n",
    "adata_raw = adata.copy()\n",
    "adata_raw.X = adata.raw.X\n",
    "adata_raw.obs['n_genes'] = (adata_raw.X > 0).sum(1) # recompute number of genes expressed per cell\n",
    "adata = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 43,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "AnnData object with n_obs × n_vars = 5197 × 13065 \n",
       "    obs: 'assignment', 'batch', 'counts', 'demux_type', 'hto_type', 'rna_type', 'sample', 'n_counts', 'log_counts', 'n_genes', 'percent_mito', 'Female', 'Female_cat', 'Female_str', 'sex_sample', 'rXist', 'leiden'\n",
       "    var: 'gene_ids', 'feature_types', 'n_cells', 'highly_variable', 'means', 'dispersions', 'dispersions_norm'\n",
       "    uns: 'leiden', 'leiden_colors', 'neighbors', 'pca', 'rank_genes_groups', 'sample_colors', 'sex_sample_colors', 'umap'\n",
       "    obsm: 'X_pca', 'X_umap'\n",
       "    varm: 'PCs'"
      ]
     },
     "execution_count": 43,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_raw"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 44,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>assignment</th>\n",
       "      <th>batch</th>\n",
       "      <th>counts</th>\n",
       "      <th>demux_type</th>\n",
       "      <th>hto_type</th>\n",
       "      <th>rna_type</th>\n",
       "      <th>sample</th>\n",
       "      <th>n_counts</th>\n",
       "      <th>log_counts</th>\n",
       "      <th>n_genes</th>\n",
       "      <th>percent_mito</th>\n",
       "      <th>Female</th>\n",
       "      <th>Female_cat</th>\n",
       "      <th>Female_str</th>\n",
       "      <th>sex_sample</th>\n",
       "      <th>rXist</th>\n",
       "      <th>leiden</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>AAACGAAGTTGGACCC-0-0-0-0-0</th>\n",
       "      <td>MPP3</td>\n",
       "      <td>batch1</td>\n",
       "      <td>902.0</td>\n",
       "      <td>singlet</td>\n",
       "      <td>background</td>\n",
       "      <td>signal</td>\n",
       "      <td>ct</td>\n",
       "      <td>13505.0</td>\n",
       "      <td>9.511259</td>\n",
       "      <td>3985</td>\n",
       "      <td>0.057805</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>true</td>\n",
       "      <td>ct_true</td>\n",
       "      <td>3.218221</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAGGATGTAGTCTGT-0-0-0-0-0</th>\n",
       "      <td>MPP3</td>\n",
       "      <td>batch1</td>\n",
       "      <td>694.0</td>\n",
       "      <td>singlet</td>\n",
       "      <td>background</td>\n",
       "      <td>signal</td>\n",
       "      <td>ct</td>\n",
       "      <td>8684.0</td>\n",
       "      <td>9.070044</td>\n",
       "      <td>3147</td>\n",
       "      <td>0.059602</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>true</td>\n",
       "      <td>ct_true</td>\n",
       "      <td>3.265923</td>\n",
       "      <td>0</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAGGTATCTTCGACC-0-0-0-0-0</th>\n",
       "      <td>MPP3</td>\n",
       "      <td>batch1</td>\n",
       "      <td>3446.0</td>\n",
       "      <td>singlet</td>\n",
       "      <td>signal</td>\n",
       "      <td>signal</td>\n",
       "      <td>ct</td>\n",
       "      <td>15873.0</td>\n",
       "      <td>9.672815</td>\n",
       "      <td>4295</td>\n",
       "      <td>0.046851</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>true</td>\n",
       "      <td>ct_true</td>\n",
       "      <td>3.039035</td>\n",
       "      <td>1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AAAGTGATCTAAGGAA-0-0-0-0-0</th>\n",
       "      <td>MPP3</td>\n",
       "      <td>batch1</td>\n",
       "      <td>783.0</td>\n",
       "      <td>singlet</td>\n",
       "      <td>background</td>\n",
       "      <td>signal</td>\n",
       "      <td>ct</td>\n",
       "      <td>16486.0</td>\n",
       "      <td>9.710449</td>\n",
       "      <td>4116</td>\n",
       "      <td>0.034508</td>\n",
       "      <td>False</td>\n",
       "      <td>False</td>\n",
       "      <td>false</td>\n",
       "      <td>ct_false</td>\n",
       "      <td>0.028160</td>\n",
       "      <td>2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>AACAAAGTCGGCTGTG-0-0-0-0-0</th>\n",
       "      <td>MPP3</td>\n",
       "      <td>batch1</td>\n",
       "      <td>951.0</td>\n",
       "      <td>singlet</td>\n",
       "      <td>background</td>\n",
       "      <td>signal</td>\n",
       "      <td>ct</td>\n",
       "      <td>17801.0</td>\n",
       "      <td>9.787459</td>\n",
       "      <td>4633</td>\n",
       "      <td>0.057387</td>\n",
       "      <td>True</td>\n",
       "      <td>True</td>\n",
       "      <td>true</td>\n",
       "      <td>ct_true</td>\n",
       "      <td>3.063702</td>\n",
       "      <td>3</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                           assignment   batch  counts demux_type    hto_type  \\\n",
       "AAACGAAGTTGGACCC-0-0-0-0-0       MPP3  batch1   902.0    singlet  background   \n",
       "AAAGGATGTAGTCTGT-0-0-0-0-0       MPP3  batch1   694.0    singlet  background   \n",
       "AAAGGTATCTTCGACC-0-0-0-0-0       MPP3  batch1  3446.0    singlet      signal   \n",
       "AAAGTGATCTAAGGAA-0-0-0-0-0       MPP3  batch1   783.0    singlet  background   \n",
       "AACAAAGTCGGCTGTG-0-0-0-0-0       MPP3  batch1   951.0    singlet  background   \n",
       "\n",
       "                           rna_type sample  n_counts  log_counts  n_genes  \\\n",
       "AAACGAAGTTGGACCC-0-0-0-0-0   signal     ct   13505.0    9.511259     3985   \n",
       "AAAGGATGTAGTCTGT-0-0-0-0-0   signal     ct    8684.0    9.070044     3147   \n",
       "AAAGGTATCTTCGACC-0-0-0-0-0   signal     ct   15873.0    9.672815     4295   \n",
       "AAAGTGATCTAAGGAA-0-0-0-0-0   signal     ct   16486.0    9.710449     4116   \n",
       "AACAAAGTCGGCTGTG-0-0-0-0-0   signal     ct   17801.0    9.787459     4633   \n",
       "\n",
       "                            percent_mito  Female Female_cat Female_str  \\\n",
       "AAACGAAGTTGGACCC-0-0-0-0-0      0.057805    True       True       true   \n",
       "AAAGGATGTAGTCTGT-0-0-0-0-0      0.059602    True       True       true   \n",
       "AAAGGTATCTTCGACC-0-0-0-0-0      0.046851    True       True       true   \n",
       "AAAGTGATCTAAGGAA-0-0-0-0-0      0.034508   False      False      false   \n",
       "AACAAAGTCGGCTGTG-0-0-0-0-0      0.057387    True       True       true   \n",
       "\n",
       "                           sex_sample     rXist leiden  \n",
       "AAACGAAGTTGGACCC-0-0-0-0-0    ct_true  3.218221      2  \n",
       "AAAGGATGTAGTCTGT-0-0-0-0-0    ct_true  3.265923      0  \n",
       "AAAGGTATCTTCGACC-0-0-0-0-0    ct_true  3.039035      1  \n",
       "AAAGTGATCTAAGGAA-0-0-0-0-0   ct_false  0.028160      2  \n",
       "AACAAAGTCGGCTGTG-0-0-0-0-0    ct_true  3.063702      3  "
      ]
     },
     "execution_count": 44,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "adata_raw.obs.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Select genes expressed in >5% of cells (no adaptive thresholding)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -i adata_raw\n",
    "\n",
    "#Convert SingleCellExperiment to SingleCellAssay type as required by MAST\n",
    "sca <- SceToSingleCellAssay(adata_raw, class = \"SingleCellAssay\")\n",
    "\n",
    "#Scale Gene detection rate\n",
    "colData(sca)$n_genes = scale(colData(sca)$n_genes)\n",
    "\n",
    "# filter genes based on hard cutoff (have to be expressed in at least 5% of all cells)\n",
    "freq_expressed <- 0.05\n",
    "expressed_genes <- freq(sca) > freq_expressed\n",
    "sca <- sca[expressed_genes,]\n",
    "\n",
    "#rename the sample to condition and make the ct the control\n",
    "cond<-factor(colData(sca)$sample)\n",
    "cond<-relevel(cond,\"ct\")\n",
    "colData(sca)$condition<-cond"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### run MAST\n",
    "\n",
    "background:  \n",
    "`zlmCond_all <- zlm(formula = ~condition + Female + n_genes, sca=sca) # this runs the model`\n",
    "\n",
    "a formula with the measurement variable (gene expression) on the LHS (left hand side) and \n",
    "predictors present in colData on the RHS\n",
    "expression of genes controlling for cluster, condition, sex + n_genes\n",
    "questions I can ask:\n",
    "sex differences controlling for treatments\n",
    "sex differences controlling for clusters - not necessary analyze all the clusters\n",
    "overall gene expression changes in treatment\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 46,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R \n",
    "#Define & run hurdle model \n",
    "zlmCond_all <- zlm(formula = ~condition + Female + n_genes, sca=sca) # this runs the model\n",
    "summaryCond_all <- summary(zlmCond_all, doLRT=TRUE) # extracts the data, gives datatable with summary of fit, doLRT=TRUE extracts likelihood ratio test p-value\n",
    "summaryDt_all <- summaryCond_all$datatable # reformats into a table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 47,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "       primerid component        contrast    Pr..Chisq.        ci.hi\n",
      "1 0610009B22Rik         C      FemaleTRUE  6.424882e-01  0.012041779\n",
      "2 0610009B22Rik         C   conditionGCSF  1.373122e-01  0.006013753\n",
      "3 0610009B22Rik         C conditiondmPGE2  2.826095e-03 -0.013251214\n",
      "4 0610009B22Rik         C   conditionindo  6.941992e-01  0.019368688\n",
      "5 0610009B22Rik         C    conditionpIC  2.122820e-02 -0.004128486\n",
      "6 0610009B22Rik         C         n_genes 3.607562e-168 -0.123017582\n",
      "        ci.lo         coef           z\n",
      "1 -0.01952166 -0.003739942  -0.4644710\n",
      "2 -0.04383080 -0.018908521  -1.4870240\n",
      "3 -0.06364322 -0.038447216  -2.9907586\n",
      "4 -0.02909560 -0.004863459  -0.3933702\n",
      "5 -0.05082359 -0.027476039  -2.3065392\n",
      "6 -0.13998017 -0.131498875 -30.3884163\n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "head(summaryDt_all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 48,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -o female_all -o GCSF_all -o dmPGE2_all -o indo_all -o pIC_all\n",
    "\n",
    "# reformat for female\n",
    "result_all_Female <- merge(summaryDt_all[contrast=='FemaleTRUE' & component=='H',.(primerid, `Pr(>Chisq)`)], #P-vals\n",
    "                  summaryDt_all[contrast=='FemaleTRUE' & component=='logFC', .(primerid, coef)],\n",
    "                  by='primerid') #logFC coefficients\n",
    "#Correct for multiple testing (FDR correction) and filtering\n",
    "result_all_Female[,FDR:=p.adjust(`Pr(>Chisq)`, 'fdr')] # create column named FDR - probably that p.adjust function\n",
    "female_all = result_all_Female[result_all_Female$FDR<0.01,, drop=F] # create new table where rows with FDR<0.01 are droped\n",
    "female_all = female_all[order(female_all$FDR),] # sorts the table\n",
    "\n",
    "\n",
    "# reformat for GCSF\n",
    "result_all_GCSF <- merge(summaryDt_all[contrast=='conditionGCSF' & component=='H',.(primerid, `Pr(>Chisq)`)], #P-vals\n",
    "                  summaryDt_all[contrast=='conditionGCSF' & component=='logFC', .(primerid, coef)],\n",
    "                  by='primerid') #logFC coefficients\n",
    "#Correct for multiple testing (FDR correction) and filtering\n",
    "result_all_GCSF[,FDR:=p.adjust(`Pr(>Chisq)`, 'fdr')] # create column named FDR - probably that p.adjust function\n",
    "GCSF_all = result_all_GCSF[result_all_GCSF$FDR<0.01,, drop=F] # create new table where rows with FDR<0.01 are droped\n",
    "GCSF_all = GCSF_all[order(GCSF_all$FDR),] # sorts the table\n",
    "\n",
    "\n",
    "# reformat for dmPGE2\n",
    "result_all_dmPGE2 <- merge(summaryDt_all[contrast=='conditiondmPGE2' & component=='H',.(primerid, `Pr(>Chisq)`)], #P-vals\n",
    "                  summaryDt_all[contrast=='conditiondmPGE2' & component=='logFC', .(primerid, coef)],\n",
    "                  by='primerid') #logFC coefficients\n",
    "#Correct for multiple testing (FDR correction) and filtering\n",
    "result_all_dmPGE2[,FDR:=p.adjust(`Pr(>Chisq)`, 'fdr')] # create column named FDR - probably that p.adjust function\n",
    "dmPGE2_all = result_all_dmPGE2[result_all_dmPGE2$FDR<0.01,, drop=F] # create new table where rows with FDR<0.01 are droped\n",
    "dmPGE2_all = dmPGE2_all[order(dmPGE2_all$FDR),] # sorts the table\n",
    "\n",
    "\n",
    "# reformat for indo\n",
    "result_all_indo <- merge(summaryDt_all[contrast=='conditionindo' & component=='H',.(primerid, `Pr(>Chisq)`)], #P-vals\n",
    "                  summaryDt_all[contrast=='conditionindo' & component=='logFC', .(primerid, coef)],\n",
    "                  by='primerid') #logFC coefficients\n",
    "#Correct for multiple testing (FDR correction) and filtering\n",
    "result_all_indo[,FDR:=p.adjust(`Pr(>Chisq)`, 'fdr')] # create column named FDR - probably that p.adjust function\n",
    "indo_all = result_all_indo[result_all_indo$FDR<0.01,, drop=F] # create new table where rows with FDR<0.01 are droped\n",
    "indo_all = indo_all[order(indo_all$FDR),] # sorts the table\n",
    "\n",
    "# reformat for pIC\n",
    "result_all_pIC <- merge(summaryDt_all[contrast=='conditionpIC' & component=='H',.(primerid, `Pr(>Chisq)`)], #P-vals\n",
    "                  summaryDt_all[contrast=='conditionpIC' & component=='logFC', .(primerid, coef)],\n",
    "                  by='primerid') #logFC coefficients\n",
    "#Correct for multiple testing (FDR correction) and filtering\n",
    "result_all_pIC[,FDR:=p.adjust(`Pr(>Chisq)`, 'fdr')] # create column named FDR - probably that p.adjust function\n",
    "pIC_all = result_all_pIC[result_all_pIC$FDR<0.01,, drop=F] # create new table where rows with FDR<0.01 are droped\n",
    "pIC_all = pIC_all[order(pIC_all$FDR),] # sorts the table"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 49,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R -o MAST_raw_all\n",
    "\n",
    "MAST_raw_all <- summaryDt_all"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 50,
   "metadata": {},
   "outputs": [],
   "source": [
    "# save files as .csvs\n",
    "\n",
    "MAST_raw_all.to_csv('./write/MPP3_MAST_raw_surf.csv')\n",
    "female_all.to_csv('./write/MPP3_MAST_female_surf.csv')\n",
    "GCSF_all.to_csv('./write/MPP3_MAST_GCSF_surf.csv')\n",
    "pIC_all.to_csv('./write/MPP3_MAST_pIC_surf.csv')\n",
    "dmPGE2_all.to_csv('./write/MPP3_MAST_dmPGE2_surf.csv')\n",
    "indo_all.to_csv('./write/MPP3_MAST_indo_surf.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 51,
   "metadata": {},
   "outputs": [],
   "source": [
    "# remove python variables\n",
    "\n",
    "adata_raw = None"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 52,
   "metadata": {},
   "outputs": [],
   "source": [
    "%%R\n",
    "# remove R variables\n",
    "\n",
    "rm(adata_raw)\n",
    "rm(zlmCond_all)\n",
    "rm(summaryDt_all)\n",
    "rm(summaryCond_all)\n",
    "rm(MAST_raw_all)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 53,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "scanpy==1.4.5.1 anndata==0.7.1 umap==0.3.10 numpy==1.17.3 scipy==1.3.0 pandas==0.25.3 scikit-learn==0.22.2.post1 statsmodels==0.10.0 python-igraph==0.7.1 louvain==0.6.1\n",
      "\n",
      "INSTALLED VERSIONS\n",
      "------------------\n",
      "commit           : None\n",
      "python           : 3.7.3.final.0\n",
      "python-bits      : 64\n",
      "OS               : Linux\n",
      "OS-release       : 4.19.76-linuxkit\n",
      "machine          : x86_64\n",
      "processor        : x86_64\n",
      "byteorder        : little\n",
      "LC_ALL           : en_US.UTF-8\n",
      "LANG             : en_US.UTF-8\n",
      "LOCALE           : en_US.UTF-8\n",
      "\n",
      "pandas           : 0.25.3\n",
      "numpy            : 1.17.3\n",
      "pytz             : 2019.3\n",
      "dateutil         : 2.8.1\n",
      "pip              : 19.3.1\n",
      "setuptools       : 41.6.0.post20191101\n",
      "Cython           : None\n",
      "pytest           : 5.3.5\n",
      "hypothesis       : None\n",
      "sphinx           : None\n",
      "blosc            : None\n",
      "feather          : None\n",
      "xlsxwriter       : None\n",
      "lxml.etree       : None\n",
      "html5lib         : None\n",
      "pymysql          : None\n",
      "psycopg2         : None\n",
      "jinja2           : 2.10.3\n",
      "IPython          : 7.9.0\n",
      "pandas_datareader: None\n",
      "bs4              : None\n",
      "bottleneck       : None\n",
      "fastparquet      : None\n",
      "gcsfs            : None\n",
      "lxml.etree       : None\n",
      "matplotlib       : 3.2.0\n",
      "numexpr          : 2.7.1\n",
      "odfpy            : None\n",
      "openpyxl         : None\n",
      "pandas_gbq       : None\n",
      "pyarrow          : None\n",
      "pytables         : None\n",
      "s3fs             : None\n",
      "scipy            : 1.3.0\n",
      "sqlalchemy       : 1.3.10\n",
      "tables           : 3.6.1\n",
      "xarray           : None\n",
      "xlrd             : None\n",
      "xlwt             : None\n",
      "xlsxwriter       : None\n"
     ]
    }
   ],
   "source": [
    "sc.logging.print_versions()\n",
    "pd.show_versions()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 54,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "R version 3.6.1 (2019-07-05)\n",
      "Platform: x86_64-conda_cos6-linux-gnu (64-bit)\n",
      "Running under: Ubuntu 18.04.2 LTS\n",
      "\n",
      "Matrix products: default\n",
      "BLAS/LAPACK: /opt/conda/lib/libopenblasp-r0.3.7.so\n",
      "\n",
      "locale:\n",
      " [1] LC_CTYPE=en_US.UTF-8       LC_NUMERIC=C              \n",
      " [3] LC_TIME=en_US.UTF-8        LC_COLLATE=en_US.UTF-8    \n",
      " [5] LC_MONETARY=en_US.UTF-8    LC_MESSAGES=en_US.UTF-8   \n",
      " [7] LC_PAPER=en_US.UTF-8       LC_NAME=C                 \n",
      " [9] LC_ADDRESS=C               LC_TELEPHONE=C            \n",
      "[11] LC_MEASUREMENT=en_US.UTF-8 LC_IDENTIFICATION=C       \n",
      "\n",
      "attached base packages:\n",
      " [1] parallel  stats4    tools     stats     graphics  grDevices utils    \n",
      " [8] datasets  methods   base     \n",
      "\n",
      "other attached packages:\n",
      " [1] Matrix_1.2-17               MAST_1.12.0                \n",
      " [3] plyr_1.8.4                  ggplot2_3.2.1              \n",
      " [5] scran_1.14.1                SingleCellExperiment_1.8.0 \n",
      " [7] SummarizedExperiment_1.16.0 DelayedArray_0.12.0        \n",
      " [9] BiocParallel_1.20.0         matrixStats_0.55.0         \n",
      "[11] Biobase_2.46.0              GenomicRanges_1.38.0       \n",
      "[13] GenomeInfoDb_1.22.0         IRanges_2.20.0             \n",
      "[15] S4Vectors_0.24.0            BiocGenerics_0.32.0        \n",
      "\n",
      "loaded via a namespace (and not attached):\n",
      " [1] Rcpp_1.0.3               rsvd_1.0.2               locfit_1.5-9.1          \n",
      " [4] lattice_0.20-38          prettyunits_1.0.2        zeallot_0.1.0           \n",
      " [7] assertthat_0.2.1         R6_2.4.0                 backports_1.1.5         \n",
      "[10] pillar_1.4.2             progress_1.2.2           zlibbioc_1.32.0         \n",
      "[13] rlang_0.4.1              lazyeval_0.2.2           data.table_1.12.6       \n",
      "[16] irlba_2.3.3              BiocNeighbors_1.4.0      statmod_1.4.32          \n",
      "[19] stringr_1.4.0            igraph_1.2.4.2           RCurl_1.95-4.12         \n",
      "[22] munsell_0.5.0            compiler_3.6.1           vipor_0.4.5             \n",
      "[25] BiocSingular_1.2.0       pkgconfig_2.0.3          ggbeeswarm_0.6.0        \n",
      "[28] tidyselect_0.2.5         tibble_2.1.3             gridExtra_2.3           \n",
      "[31] GenomeInfoDbData_1.2.2   edgeR_3.28.0             viridisLite_0.3.0       \n",
      "[34] crayon_1.3.4             dplyr_0.8.3              withr_2.1.2             \n",
      "[37] bitops_1.0-6             grid_3.6.1               gtable_0.3.0            \n",
      "[40] magrittr_1.5             scales_1.0.0             dqrng_0.2.1             \n",
      "[43] stringi_1.4.3            reshape2_1.4.3           XVector_0.26.0          \n",
      "[46] viridis_0.5.1            limma_3.42.0             scater_1.14.0           \n",
      "[49] vctrs_0.2.0              DelayedMatrixStats_1.8.0 glue_1.3.1              \n",
      "[52] beeswarm_0.2.3           purrr_0.3.3              hms_0.5.2               \n",
      "[55] abind_1.4-5              colorspace_1.4-1        \n"
     ]
    }
   ],
   "source": [
    "%%R\n",
    "\n",
    "sessionInfo()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
